### ISSUE TITLE
fix(metal): bound command-buffer stalls and surface jetsam kills under memory pressure

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12958]
## Summary

Under fleet memory pressure on a 36 GiB M3 Pro, the native-Metal `fak up`
server can (a) stall with the main thread blocked in
`-[_MTLCommandBuffer waitUntilCompleted]` and produce zero tokens for minutes,
and (b) be silently killed (SIGKILL / jetsam) mid-generation. Neither surfaces a
typed error to the client; the request just hangs then the connection drops.

## Witness (this machine, 2026-09-12)

Case A - command-buffer stall:

- `fak up --headless --context 8192` (installed build).
- `fak macobs`: `device_utilization_pct=25-26`, `free=0.13 GiB`,
  `swap_used=13.25 GiB`, `compressed=17.36 GiB`.
- `sample <pid>` main thread: `2387 -[_MTLCommandBuffer waitUntilCompleted]`
  with `IOGPUCommandQueueSubmitCommandBuffers` pending on the Metal queue.
- `POST /v1/chat/completions` (`max_tokens=4`) -> `http=000` after 30-120 s,
  **zero tokens**, process state `U` (uninterruptible wait).

Case B - silent kill:

- `fak up --headless --context 20000` (trunk c5f748095, cgo, Q8 residency on,
  `metal_live_q8_weights=272`).
- A `max_tokens=8` request ran >5 min at 100-330% CPU (active in
  `prefillQwen35HybridQ4KLayer` / `kQuantGemmDispatch`, no panic), then the
  process vanished with no panic/fatal line in `/tmp/fak-goal3.log` and the
  client saw `http=000` at 335 s. System free pages had fallen to ~90 MB.

## Why this matters for the goal

Even with Q8 GPU residency and 20k context landed, the box cannot meet
`>30 prefill / >7 decode` while co-resident with a fleet: the Metal queue stalls
or the process is killed under the 36 GiB ceiling. The failure is **untyped and
unbounded** - no timeout, no retry, no surfaced reason - so "out of the box"
degrades to a hang rather than a clear refusal.

## Proposed direction

1. Bound the Metal dispatch: a watchdog/timeout on command-buffer completion
   that converts a stall into a typed `MetalCommandBufferStalled` error the
   server can return (HTTP 503 + reason), never an unbounded wait.
2. Surface the eviction/preemption signal: record `recovery_count`,
   `memorystatus_level`, and the wired-limit headroom at stall time so the
   receipt explains the kill.
3. Prefer the headroom admission (#12957) so the process is not admitted into a
   state where the GPU queue will be preempted; reserve activation headroom
   instead of relying on the live-free snapshot.

## Evidence refs

- Live `sample` stack: `_MTLCommandBuffer waitUntilCompleted` /
  `IOGPUCommandQueueSubmitCommandBuffers`.
- `/tmp/fak-goal3.log` (clean load, `metal_live_q8_weights=272`, abrupt end, 0
  panic lines).
- Related: #12957 (memory-pressure context collapse), #12955 (Q2_K Metal panic).

**Comment by anthony-chaudhary** [comment]
## Landed (witnessed, pushed)

Commit `11130e01b` on `origin/main` — `fix(metal): bound command-buffer waits with a typed stall and eviction headroom receipt (#12958)`:

- **`internal/metalgemm/metal_stall.go`** — typed `MetalCommandBufferStallError{Operation, WaitedMilliseconds, LimitMilliseconds}`, `IsMetalCommandBufferStall` (via `errors.As`), `DefaultCommandBufferWaitLimit`, and `CheckCommandBufferWait(op, waitedMS, limitMS)`. Contract is total: healthy only for a finite, non-negative wait strictly under a finite, positive limit; inclusive at the limit; a NaN/±Inf/negative wait or invalid limit is a stall (an unmeasurable wait is never reported healthy).
- **`internal/macobs/eviction_pressure.go`** — `EvictionPressure{RecoveryCount, MemorystatusLevel, WiredLimitBytes, ResidentBytes, HeadroomBytes, HeadroomFraction, AtRisk}`, pure `BuildEvictionPressure(...)` and `IsEvictionRisk(memorystatusLevel, headroomBytes, reserveBytes)`. Reuses the existing `HardwareTelemetry.RecoveryCount` / `MemoryStatusLevel` darwin sysctl seam; uint64 underflow guarded (resident > wired ⇒ 0 headroom); unknown wired limit ⇒ fraction 0, abstains rather than misreading missing data as pressure.

Witness (on the committed content):
```
ok  internal/metalgemm  0.21s   (14 stall subtests PASS)
ok  internal/macobs     0.27s   (BuildEvictionPressure 8 + IsEvictionRisk 6 PASS)
gofmt -l → clean
```
Independently cross-validated: both units are mutation-sensitive (deleting the impl flips an assertion); the underflow guard and zero-wired abstain branches are each covered by a would-fail-on-delete assertion.

## Verified but blocked from landing: HTTP surfacing (Lane C)

The error→HTTP mapping is written and green, but cannot land right now because the two target files are claimed by **unlanded sibling deliverables** (the safecommit `PEER_WIP_COLLISION` guard refuses, correctly):

- `internal/gateway/http.go` — claimed by parked WIP `refs/fak/wip/gateway-recovery-4c58` (dead session, `CLOSED_DIRTY_RECOVERABLE`: `turn_cost.go` et al. not yet landed).
- `cmd/fak/up.go` — its isolated build also needs the still-untracked `metalResidencyStampFrom` / `RefreshMetalResidency` peer symbols.

The verified wiring is preserved in `refs/fak/wip/metal-stall-12958-laneC` (buildable=true). It is a two-arm addition, independently witnessed green:
- `internal/gateway/http.go` `upstreamErrorStatus`: `errors.As` → `(503, "metal_command_buffer_stalled", msg naming op + waited/limit ms)`.
- `cmd/fak/up.go` `writeTurnkeyInferenceError`: same → 503 with the `metal_command_buffer_stalled` JSON code (currently a generic 500).
- Witness: `TestUpstreamErrorStatus_MetalCommandBufferStall` PASSES with a `%w`-wrapped error, proving `errors.As` walks the chain.

## Not yet done (honest remainder)

1. **Bound the actual wait at the source.** `CheckCommandBufferWait` is a bounded *classifier*, but the native `.m` waits (`[cmd waitUntilCompleted]` in q8.m/q4k.m/q2_0.m/decode.m/…) still block unboundedly; wiring the observation to a real timeout (e.g. `waitUntilCompleted` → `addCompletedHandler` + timed fallback, or a watchdog thread) is the remaining kernel half.
2. Land Lane C once the sibling deliverables clear (recover from `refs/fak/wip/metal-stall-12958-laneC`).
3. Feed `macobs.EvictionPressure` into the `fak up`/`fak serve` receipt so a jetsam kill records `recovery_count` / `memorystatus_level` / wired headroom (today `fak up` wires no memory governor).

Related: #12957 (wired-limit headroom admission).

**Comment by anthony-chaudhary** [comment]
## In-kernel `fak up` (turnkey/graph) half now landed — independent of the gateway half

The typed-stall spine from this issue (`internal/metalgemm/metal_stall.go`) is now
**wired into production on the in-kernel turnkey path**, which was a distinct
unbounded-wait site from the gateway path in #13050.

**Commit:** `91e91ded2` on public `fak` main — `fix(metalgemm): classify a
non-completed graph command buffer as a typed stall (fak metalgemm)`.

### Witness (live, this 36 GiB M3 Pro, 2026-09-18)
- A live `fak-native up --headless --model 27B --context 20480` hung forever on
  `POST /v1/chat/completions` (curl `-m 300` -> HTTP 000; urllib 600s timeout).
- `sample <pid>` placed the request goroutine at:
  `mg_graph_finish -> -[MTLCommandBuffer waitUntilCompleted] -> __psynch_cvwait`
  (`internal/metalgemm/q4k.m:2064`) — precisely the Case A stall this issue names.
- Trigger = memory exhaustion: `memory_pressure` reported 53.6% compressed and
  ~4.9 GiB allocatable. The same 27B model completed correctly once memory was
  freed, so the command buffer never reaching `Completed` was a pressure-induced
  device fault, not a bad shader.
- A SIGKILL'd wedged process sat in state `?E` for ~40s, unkillable during Metal
  driver teardown — consistent with the jetsam/teardown half of this issue.

### What landed
- `mg_graph_finish` records `status_code`, `error_code`, `device_ok`, and a
  bounded `error_text[256]` into the receipt; `inject=2` is a device-fault test seam.
- `ProjectionGraph.Finish()` routes a committed-but-not-completed buffer to
  `CheckCommandBufferWait`, surfacing typed `MetalCommandBufferStallError`.
- New `graph_finish_stall_darwin_test.go`; fails before, passes after.
- Fixed-binary E2E: 27B request returned in 54.8s (`prefill=0.6tok/s`,
  `decode=2.4tok/s`); 4B returned in 10.76s (`prefill=5.3tok/s`, `decode=3.7tok/s`).

### Still open on this issue
- The error->HTTP surfacing for the in-kernel `fak up` server (the gateway half
  is #13050). The in-kernel handler currently gets `nil`-or-error from
  `Finish()`; mapping the typed stall to a clean HTTP status + client-visible
  body on the turnkey path is not yet done.
- The RUNNING server still has no memory-pressure feedback: the admission check
  refuses on the *next* boot (correct, fail-closed) but a server already loaded
  has no live pressure signal before it faults.

**Comment by anthony-chaudhary** [comment]
**Witnessed 2026-09-18 (Apple M3 Pro, fak trunk c1c9fb8bf) — one seam of this class is now bounded; the Q4_K half of the same class remains.**

Live hang reproduced on `fak-native up --headless --model 27B`: a `sample` of the hung process showed one thread 100% in `mg_q8_gemv -> -[_MTLCommandBuffer waitUntilCompleted]`, with the matching `com.Metal.CommandQueueDispatch` thread 100% in `-[IOGPUMetalCommandQueue _submitCommandBuffers:count:]` — the GPU submit never returned. `/v1/chat/completions` never answered.

**Fixed (landed c1c9fb8bf):** both Q8 GEMV waits in `internal/metalgemm/q8.m` are now bounded.
- `[cmd waitUntilCompleted]` replaced with a `dispatch_semaphore` completed-handler wait bounded to 10s (parity with `DefaultCommandBufferWaitLimit`), handler registered BEFORE commit.
- On timeout: `completed_wait=0`, `*out_status=-1`, `*out_error=ETIMEDOUT`, and the function returns WITHOUT reading unwritten output or re-entering `waitUntilCompleted`.
- The typed stall is threaded into the live serve path: `internal/model/metal_q4k_on.go` `q4kGroupDispatch` uses `GEMVGroupQ8WithEventsErr` and declines on `IsMetalCommandBufferStall`, so `mulGroup` falls back to the proven per-weight path and the turn still completes.
- Witness: `go test ./internal/metalgemm/ -count=1` → ok; 5 `TestQ8GEMV*` tests PASS on the real Metal device (none skipped). Live: after the fix the same server answered 2/12/25-token prompts in 3.2-3.5s.

**Still open in this issue's class:** `internal/metalgemm/q4k.m` still calls the unbounded `[cmd waitUntilCompleted]` at ~6 sites (`mg_q4k_mlp:924`, `mg_q6k_gemv:1055`, `mg_q6k_gemm:1092`, `mg_q4k_mlp_q6down_batch:1215`, `mg_q4k_mlp_q6down:1287`, plus ~1413). Those are the same defect shape and are NOT yet bounded — the Q4_K decode/prefill path can still block unbounded on a wedged GPU.

Also relevant: `fak#13050` (surface the stall as HTTP 503) — today the stall silently degrades to the CPU path, so the client sees a slow-but-successful turn rather than a diagnosable failure.

**Comment by anthony-chaudhary** [comment]
## 2026-09-20 M3 Pro route-survival diagnostic: retained, not performance-qualified

A source-pinned `modelbench` P40/T16 diagnostic with the automatic Qwen3.8 resident whole-sequence selector disabled completed instead of hitting the graph-finish 10s stall:

- load: 23.691s
- prefill: 40 tokens / 9.321s = 4.291 tok/s
- decode: 16 steps, 587.055ms/token = 1.703 tok/s
- process: exit 0, bounded under 210s and 24 GiB sampled RSS

This result **does not isolate the selector** and must not be used as a performance comparison. The diagnostic harness differed materially from the live `fak up` runtime:

1. It expanded the Q4_K token embedding to F32 (`f32_bytes=5,096,175,616`, `q4k_embed_tensors=0`, total reported residency `24,323,393,536`). The server selects packed Q4_K embedding residency (`f32_bytes=10,582,016`, `q4k_embed_tensors=1`, total `19,952,961,536`), saving `4,370,432,000` bytes.
2. The harness report names Q8 minority execution on CPU and did not call `PrepareMetalResidency`; the live server prepares Metal residency and reports 272 Q8 plus 41 Q6 weights resident.
3. Memory was under severe pressure during the run: sampled RSS peaked at `25,671,172,096` bytes, about 98.6 MiB below the bound; swap grew from 7.45 GiB to 11.48 GiB during inference and system free memory reached 16%. Swap later fell after teardown/restoration.

The raw diagnostic is retained as a route-survival artifact, labeled polluted and not performance-qualified. No selector-ON rerun or 20k run was attempted.

The next valid physical comparison must first make the benchmark recipe match the server: select the eligible packed Qwen3.8 embedding, eagerly prepare Metal residency before all timed sessions, stamp both choices in the report, then run in a fresh process only after memory pressure and swap-growth admission are clean. The 27B Q4_K_M comparison remains separate from a possible 27B UD-Q2_K_XL envelope; there is no smaller Qwen3.8 artifact cached on this machine and no exact-machine alternate qualification yet.

**Comment by anthony-chaudhary** [comment]
## 2026-09-20 source-pinned production-path rerun: memory gate aborts before headers

A clean source-pinned `fak up` rerun used the exact production loader/session, the same tokenizer-confirmed P40/T16 SSE request, and a quiescent post-service admission window. The static fit plan admitted 27B Q4_K_M + fp32 20,480 context with authored 41.7% headroom. Post-bootout live admission observed 91–92% free memory and stable 1,648 MiB swap.

The production load completed in 17.916s and reported the intended recipe: packed Q4_K embedding, 272 Q8 + 41 Q6 Metal-resident weights, total model residency 19,952,961,536 bytes. During load, swap rose by 95 MiB. Immediately after the 40-token request was dispatched, swap reached 3,816 MiB (a further ~2.1 GiB) and free memory fell to 34%. The reviewed safety monitor aborted after two consecutive samples exceeded the +64 MiB swap-growth threshold. Result: curl exit 52, zero SSE bytes, no response headers; the original launchd service was restored healthy.

This run did not reach the graph's 10s finish timeout. It establishes an earlier physical failure boundary: the production whole-sequence P40 request creates enough transient pressure to force sustained swap growth even from a quiescent admission state. `NewKVCache` is lazy and grows with actual rows, so this tiny request did not reserve the full 20,480-token KV budget. The current leading allocation source is the one-command-buffer 64-layer `ProjectionGraph`, which retains every intermediate `GraphResult` buffer until terminal finish/free.

Forward work:

1. include the whole-sequence graph transient working set in native admission/receipts;
2. bound graph live memory through layer segmentation with device-resident hidden/state handoff, or a separately proven in-command-buffer liveness reuse scheme;
3. keep timed-out GDN owners quarantined until actual command-buffer completion before enabling HTTP recovery.

No 20k run or retry followed this safety abort.

**Comment by anthony-chaudhary** [comment]
## Claimed bounded implementation slice: prefill graph live-set reuse

The physical P40 production-path receipt localizes the first actionable allocation defect: the prefill `Qwen35MetalForwardSequence` retains all intermediate graph buffers across 64 layers. The existing P1 decode graph already bounds this same class through `ProjectionGraph.SetBufferPool(8)` plus last-consumer `Release` calls, and the Metal primitive has an on-device parity test.

This slice is limited to `internal/model/metal_prefill_hybrid.go` and one focused model test. It will enable the existing pool only on the already-qualified Qwen3.8 Metal sequence route and mirror the decode graph's last-consumer map. Terminal hidden/KV/raw-capture results remain retained. Eligibility, kernels, command-buffer count, waits, readbacks, context semantics, and fallback rules remain unchanged.

Acceptance is staged:

1. synthetic physical-Metal OFF/ON test: distinct equivalent sessions, bit-identical logits/KV/GDN state, unchanged receipt topology, no fallback;
2. independently reviewed source-pinned P40/T16 production SSE microdose under the existing RSS/free/swap bounds;
3. default remains limited to the current qualified route, and larger/full-model claims wait for physical memory and throughput evidence.

This coordinates with #11582. It does not overlap the Metal command-buffer/GDN quarantine work in `internal/metalgemm`.
